aws/resource_aws_appsync_graphql_api.go

```diff
@@ -0,0 +1,215 @@
+								validTypes := []string{"ALLOW", "DENY"}
+								for _, str := range validTypes {
+									if value == str {
+										return
+									}
+								}
+								errors = append(errors, fmt.Errorf("expected %s to be one of %v, got %s", k, validTypes, value))
+								return
+							},
+							StateFunc: func(v interface{}) string {
+								return strings.ToUpper(v.(string))
+							},
+						},
+						"user_pool_id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppsyncGraphqlApiCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.CreateGraphqlApiInput{
+		AuthenticationType: aws.String(d.Get("authentication_type").(string)),
+		Name:               aws.String(d.Get("name").(string)),
+		UserPoolConfig:     expandAppsyncGraphqlApiUserPoolConfig(d.Get("user_pool_config").([]interface{})),
+	}
+
+	resp, err := conn.CreateGraphqlApi(input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*resp.GraphqlApi.ApiId)
+	d.Set("arn", resp.GraphqlApi.Arn)
+	return nil
+}
+
+func resourceAwsAppsyncGraphqlApiRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.GetGraphqlApiInput{
+		ApiId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetGraphqlApi(input)
+	if err != nil {
+		if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("authentication_type", resp.GraphqlApi.AuthenticationType)
+	d.Set("name", resp.GraphqlApi.Name)
+	d.Set("user_pool_config", flattenAppsyncGraphqlApiUserPoolConfig(resp.GraphqlApi.UserPoolConfig))
+	d.Set("arn", resp.GraphqlApi.Arn)
+	return nil
+}
+
+func resourceAwsAppsyncGraphqlApiUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.UpdateGraphqlApiInput{
+		ApiId: aws.String(d.Id()),
+		Name:  aws.String(d.Get("name").(string)),
+	}
+
+	if d.HasChange("authentication_type") {
+		input.AuthenticationType = aws.String(d.Get("authentication_type").(string))
+	}
+	if d.HasChange("user_pool_config") {
+		input.UserPoolConfig = expandAppsyncGraphqlApiUserPoolConfig(d.Get("user_pool_config").([]interface{}))
+	}
+
+	_, err := conn.UpdateGraphqlApi(input)
+	if err != nil {
+		if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	return resourceAwsAppsyncGraphqlApiRead(d, meta)
+}
+
+func resourceAwsAppsyncGraphqlApiDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.DeleteGraphqlApiInput{
+		ApiId: aws.String(d.Id()),
+	}
+	_, err := conn.DeleteGraphqlApi(input)
+	if err != nil {
+		if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+			d.SetId("")
```

Comment:
Nitpick: I know we still have it in many resources, but it's redundant. The ID is emptied automatically in the core.

See https://github.com/hashicorp/terraform/blob/096847c9f774bfb946de7413260b30f9f6041241/helper/schema/resource.go#L197-L206